File: android/sdk/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/ContactChannelStrategy.java
package org.elastos.sdk.elephantwallet.contact.internal;

import org.elastos.sdk.elephantwallet.contact.Contact;
import org.elastos.tools.crosspl.CrossBase;
import org.elastos.tools.crosspl.annotation.CrossClass;
import org.elastos.tools.crosspl.annotation.CrossInterface;

@CrossClass
abstract class ContactChannelStrategy extends CrossBase {
    protected ContactChannelStrategy(int channelId, String name) {
        super(ContactChannelStrategy.class.getName(), 0);
        mChannel = new Contact.Channel(channelId, name);
        int ret = syncChannelToNative(channelId, name);
        if(ret < 0) {
            throw new RuntimeException("Failed to sync channel to native.");
        }
    }

    public Contact.Channel getChannel() {
        return mChannel;
    }

    public int getChannelId() {
        return mChannel.id();
    }

    public abstract int onSendMessage(String humanCode, byte[] data);

    public int receivedMessage(String humanCode, byte[] data) {
        int ret = receivedMessage(humanCode, mChannel.id(), data);
        return ret;
    }

    @CrossInterface
    public int onOpen() {
        return 0;
    }

    @CrossInterface
    public int onClose() {
        return 0;
    }

    @CrossInterface
    public int onSendMessage(String humanCode, int channelId, byte[] data) {
        assert(channelId == mChannel.id());

        int ret = onSendMessage(humanCode, data);
        return ret;
    }

    @CrossInterface
    public native int receivedMessage(String humanCode, int channelId, byte[] data);

    @CrossInterface
    public native int syncChannelToNative(int channelId, String name);

    Contact.Channel mChannel;
}

File: ios/sdk/swift/internal/ContactChannelStrategy.swift
import CrossPL

/* @CrossClass */
@objc open class ContactChannelStrategy: CrossBase {
  public init(channelId: Int, name: String) {
    super.init(className: String(describing: ContactChannelStrategy.self))
    mChannel = Contact.Channel(rawValue: channelId)
    let ret = syncChannelToNative(channelId, name)
    if(ret < 0) {
      fatalError("Failed to sync channel to native.");
    }
  }
  
  public func getChannel() -> Contact.Channel {
      return mChannel!;
  }
  
  public func getChannelId() -> Int {
      return mChannel!.rawValue;
  }

  open func onSendMessage(humanCode: String, data: Data?) -> Int {
    fatalError("\(#function) not implementation.")
  }

  public func receivedMessage(humanCode: String, data: Data?) -> Int {
    let ret = receivedMessage(humanCode, mChannel!.rawValue, data);
    return ret;
  }

  /* @CrossPlatformInterface */
  @objc open func onOpen() -> Int {
    return 0
  }
  
  /* @CrossPlatformInterface */
  @objc open func onClose() -> Int {
    return 0
  }
  
  /* @CrossPlatformInterface */
  @objc internal func onSendMessage(_ humanCode: String, _ channelType: Int, _ data: Data?) -> Int {
    let ret = onSendMessage(humanCode: humanCode, data: data)
    return ret
  }
  
  /* @CrossNativeInterface */
  private func receivedMessage(_ humanCode: String, _ channelType: Int, _ data: Data?) -> Int {
    let ret = crosspl_Proxy_ContactChannelStrategy_receivedMessage(nativeHandle,
                                                                   humanCode, Int32(channelType), data)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func syncChannelToNative(_ channelId: Int, _ name: String) -> Int {
    let ret = crosspl_Proxy_ContactChannelStrategy_syncChannelToNative(nativeHandle,
                                                                       Int32(channelId), name)
    return Int(ret)
  }
  
  private var mChannel: Contact.Channel? = nil;
}
